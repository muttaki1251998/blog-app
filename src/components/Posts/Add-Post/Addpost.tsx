import React from "react";
import axios from "axios";
import * as Yup from "yup";
import { Formik, Form, Field } from "formik";

const Addpost = () => {
  const handleSubmit = async (values: any) => {
    try {
      const res = await axios.post("http://localhost:8000/api/add", values);
      console.log("Post added", res);
    } catch (e) {
      console.log("Unable to add post", e);
    }
  };

  const validationSchema = Yup.object().shape({
    title: Yup.string()
      .required("Required")
      .min(4, "Too Short!")
      .max(15, "Too Long!"),
    description: Yup.string()
      .required("Required")
      .min(10, "Too Short!")
      .max(50, "Too Long!"),
  });

  const initialValues = {
    title: "",
    description: "",
  };

  return (
    <Formik
      initialValues={initialValues}
      validationSchema={validationSchema}
      onSubmit={handleSubmit}
    >
      {({ errors, touched }) => (
        <Form
          className="flex items-center justify-center"
          style={{ height: "calc(100vh - 64px)", overflowY: "auto" }}
        >
          <div className="w-full max-w-md">
            <label className="text-lg text-gray-800 tracking-wide text-center block mb-4">
              Create a post
            </label>
            <div className="mb-3">
              <Field
                type="text"
                name="title"
                data-testid="title"
                placeholder="Title"
                className={`shadow appearance-none border ${
                  errors.title && touched.title
                    ? "border-red-500"
                    : "border-gray-200"
                } rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-blue-100`}
              />
              {errors.title && touched.title && (
                <div className="text-red-500 text-xs mt-1">{errors.title}</div>
              )}
            </div>
            <div className="mb-3">
              <Field
                as="textarea"
                data-testid="description"
                name="description"
                placeholder="Description"
                className={`shadow appearance-none border ${
                  errors.description && touched.description
                    ? "border-red-500"
                    : "border-gray-200"
                } rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-blue-100 h-32`}
              />
              {errors.description && touched.description && (
                <div className="text-red-500 text-xs mt-1">
                  {errors.description}
                </div>
              )}
            </div>
            <button
              type="submit"
              data-testid="submit-btn"
              disabled={errors.title || errors.description ? true : false}
              className={`bg-black text-white rounded py-2 px-4 hover:bg-gray-600 transition duration-300 ease-in-out ${
                errors.title || errors.description
                  ? "opacity-50 cursor-not-allowed"
                  : "hover:bg-gray-600"
              }`}
            >
              Create
            </button>
          </div>
        </Form>
      )}
    </Formik>
  );
};

export default Addpost;
