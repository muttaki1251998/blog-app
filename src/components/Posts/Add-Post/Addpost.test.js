import React from "react";
import "@testing-library/jest-dom";
import { act } from "react-dom/test-utils";
import { render, screen, fireEvent, getByTestId } from "@testing-library/react";
import Addpost from "./Addpost";

describe("snapshot", () => {
  it("matches the snapshot", () => {
    const { asFragment } = render(<Addpost />);
    expect(asFragment()).toMatchSnapshot();
  });
});
describe("Addpost Component", () => {
  test("initial UI is rendered as expected", () => {
    render(<Addpost />);
    expect(screen.getByPlaceholderText("Title")).toBeInTheDocument();
    expect(screen.getByPlaceholderText("Description")).toBeInTheDocument();
    expect(screen.getByText("Create")).toBeInTheDocument();
  });
  it("should show errors when inputting wrong texts", async () => {

      render(<Addpost />);
      const titleInput = screen.getByTestId("title");

      fireEvent.change(titleInput, { target: { value: "Ab" } });
      fireEvent.blur(titleInput);

      expect(
        await screen.findByText("Too Short!", {}, { timeout: 500 })
      ).toBeInTheDocument();

  });
});
