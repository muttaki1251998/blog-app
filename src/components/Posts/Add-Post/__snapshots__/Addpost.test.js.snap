// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot matches the snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
    class="flex items-center justify-center"
    style="height: calc(100vh - 64px); overflow-y: auto;"
  >
    <div
      class="w-full max-w-md"
    >
      <label
        class="text-lg text-gray-800 tracking-wide text-center block mb-4"
      >
        Create a post
      </label>
      <div
        class="mb-3"
      >
        <input
          class="shadow appearance-none border border-gray-200 rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-blue-100"
          data-testid="title"
          name="title"
          placeholder="Title"
          type="text"
          value=""
        />
      </div>
      <div
        class="mb-3"
      >
        <textarea
          class="shadow appearance-none border border-gray-200 rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-blue-100 h-32"
          data-testid="description"
          name="description"
          placeholder="Description"
        />
      </div>
      <button
        class="bg-black text-white rounded py-2 px-4 hover:bg-gray-600 transition duration-300 ease-in-out hover:bg-gray-600"
        data-testid="submit-btn"
        type="submit"
      >
        Create
      </button>
    </div>
  </form>
</DocumentFragment>
`;
